### PR TITLE
feat: Add Datadog integration

### LINKED ISSUE #45: Support Datadog as Backend
Be able to use Datadog as monitoring backend
And to export error_burn_rate metric

### PR DESCRIPTION
**Pull request by ocervell**
Add support for Datadog backend and Datadog exporter.
Fixes #45 
- [x] Datadog backend
  - [x] `query_sli`
  - [x] `query_slo`
  - [x] `good_bad_ratio`
  - ~~[x] `distribution_cut`~~ (removed since DD histograms are not standard distributions, see docs/datadog.md explanation)
- [x] Datadog exporter
- [x] Docs
   - [x] Backend
   - [x] Exporter
- [x] Unittests
  - [x] Fixtures
  - [x] Tests
  - [x] DD API mock

### CODE REVIEW

File: docs/datadog.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+**Config example:**
+
+```yaml
+backend:
+  class:   Datadog
+  method:  good_bad_ratio
+  api_key: ${DATADOG_API_KEY}
+  app_key: ${DATADOG_APP_KEY}
+  measurement:
+    filter_good: sum:system.disk.used{*}
```
Comment (steventhurgood) [inline docs/datadog.md]: I'm not sure that the example of disk used / disk total is a good one for SLOs, as it's more of a cause rather than the symptom-based SLOs we usually advocate.

File: docs/providers/datadog.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    filter_good: sum:system.disk.used{*}
+    filter_valid: sum:system.disk.total{*}
+```
+**&rightarrow; [Full SLO config](../samples/datadog/slo_dd_disk_utilization_ratio.yaml)**
+
+
+### Query SLI
+
+The `query_sli` method is used to directly query the needed SLI with Datadog:
+indeed, Datadog's query language is powerful enough that it can do ratios
```
Comment (steventhurgood) [inline docs/providers/datadog.md:45]: I would drop the "indeed, " (also elsewhere) and leave it as:

The `query_sli` method is used to directly query the needed SLI with Datadog: Datadog's query language is powerful enough that it can do ratios natively

File: docs/providers/datadog.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+**Config example:**
+
+```yaml
+backend:
+  class:   Datadog
+  method:  good_bad_ratio
+  api_key: ${DATADOG_API_KEY}
+  app_key: ${DATADOG_APP_KEY}
+  measurement:
+    filter_good: sum:system.disk.used{*}
```
Comment (steventhurgood) [inline docs/providers/datadog.md:36]: I don't think disk usage is a good example for an SLO. We encourage people to create SLOs based upon symtoms: things that are actually impacting users. This feels more like it's encouraging a cause-based approach.

File: docs/providers/datadog.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  api_key: ${DATADOG_API_KEY}
+  app_key: ${DATADOG_APP_KEY}
+```
+
+**&rightarrow; [Full SLO config](../samples/datadog/slo_dd_disk_utilization_query_slo.yaml)**
+
+### Examples
+
+Complete SLO samples using `Datadog` are available in
+[samples/datadog](../samples/datadog). Check them out !
```
Comment (steventhurgood) [inline docs/providers/datadog.md:88]: No space before the !:

Check them out!

File: slo_generator/backends/datadog.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+        Args:
+            :dict: Timeseries response from Datadog Metrics API endpoint.
+
+        Returns:
+            int: Event count.
+        """
+        try:
+            pointlist = timeseries['series'][0]['pointlist']
+            values = [point[1] for point in pointlist for i in point]
```
Comment (steventhurgood) [inline slo_generator/backends/datadog.py:165]: Try and avoid multiple for clauses in the same list comprehension, as it makes it harder to read:

https://google.github.io/styleguide/pyguide.html#27-comprehensions--generator-expressions

File: docs/datadog.md (hunk 1 above)
Comment (ocervell) [reply docs/datadog.md]: yep, I agree - I had to work with the metrics I had in Datadog but will add more meaningful samples.

File: samples/prometheus/slo_prom_metrics_availability_ratio.yaml [hunk 6]
```diff
@@ -26,9 +26,9 @@ backend:
   #   Content-Type: application/json
   #   Authorization: Basic b2s6cGFzcW==  # username:password base64-encoded
   measurement:
-    filter_good:  prometheus_http_requests_total{handler="/metrics", code=~"2.."}[window]
-    filter_valid: prometheus_http_requests_total{handler="/metrics"}[window]
-    # filter_bad: prometheus_http_requests_total{code=~"5..", handler="/metrics"}[window]  # alternative to filter_valid field
+    filter_good:  prometheus_http_requests_total{handler="/metrics", code=~"2.."}
+    filter_valid: prometheus_http_requests_total{handler="/metrics"}
```
Comment (ocervell) [inline samples/prometheus/slo_prom_metrics_availability_ratio.yaml:30]: should be done in a separate PR, has no relation with Datadog

File: docs/providers/datadog.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+i.e a set of configurable bins, each providing the number of events falling into
+each bin.
+
+Standard histograms representations (see [wikipedia](https://en.wikipedia.org/wiki/Histogram))
+already implement this, but the approach Datadog took is to pre-compute
+(client-side) or post-compute (server-side) percentiles, resulting in a
+different metric for each percentile representing the percentile value instead
+of the number of events in the percentile.
+
+This implementation has a couple advantages, like making it easy to query and
```
Comment (steventhurgood) [inline docs/providers/datadog.md:134]: "has a couple advantages" -> "has a couple of advantages" or just "has advantages"

File: docs/datadog.md (hunk 1 above)
Comment (ocervell) [reply docs/datadog.md]: updated, added configs for app availability instead

File: docs/providers/datadog.md (hunk 2 above)
Comment (ocervell) [reply docs/providers/datadog.md]: Done !

File: docs/providers/datadog.md (hunk 3 above)
Comment (ocervell) [reply docs/providers/datadog.md]: Done

File: docs/providers/datadog.md (hunk 4 above)
Comment (ocervell) [reply docs/providers/datadog.md]: Fixed

File: slo_generator/backends/datadog.py (hunk 5 above)
Comment (ocervell) [reply slo_generator/backends/datadog.py]: Refactor using a standard `for` loop

File: docs/providers/datadog.md (hunk 7 above)
Comment (ocervell) [reply docs/providers/datadog.md]: Fixed

**Review by steventhurgood — APPROVED** [review]
